src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -0,0 +1,508 @@
+            status = EMBER_ZCL_STATUS_INVALID_VALUE;
+        else
+            status = EMBER_ZCL_STATUS_SUCCESS;
+
+        break;
+    }
+
+    case Ids::SystemMode: {
+        uint8_t ControlSequenceOfOperation = 0xff; // Invalid value
+        uint8_t RequestedSystemMode        = 0xff; // Invalid value
+        GetControlSequenceOfOperation(endpoint, &ControlSequenceOfOperation);
+        RequestedSystemMode = *value;
+        if (ControlSequenceOfOperation > EMBER_ZCL_THERMOSTAT_CONTROL_SEQUENCE_COOLING_AND_HEATING_WITH_REHEAT ||
+            RequestedSystemMode > EMBER_ZCL_THERMOSTAT_SYSTEM_MODE_FAN_ONLY)
+        {
+            status = EMBER_ZCL_STATUS_INVALID_VALUE;
+        }
+        else
+        {
+            switch (ControlSequenceOfOperation)
+            {
+            case EMBER_ZCL_THERMOSTAT_CONTROL_SEQUENCE_COOLING_ONLY:
+            case EMBER_ZCL_THERMOSTAT_CONTROL_SEQUENCE_COOLING_WITH_REHEAT:
+                if (RequestedSystemMode == EMBER_ZCL_THERMOSTAT_SYSTEM_MODE_HEAT ||
+                    RequestedSystemMode == EMBER_ZCL_THERMOSTAT_SYSTEM_MODE_EMERGENCY_HEATING)
+                    status = EMBER_ZCL_STATUS_INVALID_VALUE;
+                break;
+
+            case EMBER_ZCL_THERMOSTAT_CONTROL_SEQUENCE_HEATING_ONLY:
+            case EMBER_ZCL_THERMOSTAT_CONTROL_SEQUENCE_HEATING_WITH_REHEAT:
+                if (RequestedSystemMode == EMBER_ZCL_THERMOSTAT_SYSTEM_MODE_COOL ||
+                    RequestedSystemMode == EMBER_ZCL_THERMOSTAT_SYSTEM_MODE_PRECOOLING)
+                    status = EMBER_ZCL_STATUS_INVALID_VALUE;
+                break;
+            default:
+                break;
+            }
+        }
+    }
+    default:
+        break;
+    }
+
+    return status;
+}
+
+bool emberAfThermostatClusterClearWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj)
+{
+    // TODO
+    return false;
+}
+bool emberAfThermostatClusterGetRelayStatusLogCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj)
+{
+    // TODO
+    return false;
+}
+
+bool emberAfThermostatClusterGetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t daysToReturn, uint8_t modeToReturn)
+{
+    // TODO
+    return false;
+}
+
+bool emberAfThermostatClusterSetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t numberOfTransitionsForSequence, uint8_t daysOfWeekForSequence,
+                                                       uint8_t modeForSequence, uint8_t * payload)
+{
+    // TODO
+    return false;
+}
+
+using namespace chip::app::Clusters::Thermostat::Attributes;
+
+int16_t EnforceHeatingSetpointLimits(int16_t HeatingSetpoint, EndpointId endpoint)
+{
+    // Optional Mfg supplied limits
+    int16_t AbsMinHeatSetpointLimit = 700;  // 7C (44.5 F) is the default
```

Comment:
a constexpr constant would help reuse things here.

Reply:
Done in all places for all magic numbers.